networkit/independentset.pyx

```diff
@@ -6,31 +6,84 @@ from libcpp.string cimport string
 
 from .graph cimport _Graph, Graph
 
-# TODO: expose all methods
-cdef extern from "<networkit/independentset/Luby.hpp>":
+cdef extern from "<networkit/independentset/IndependentSetFinder.hpp>":
 
-	cdef cppclass _Luby "NetworKit::Luby":
-		_Luby() except +
-		vector[bool_t] run(_Graph G) except +
+	cdef cppclass _IndependentSetFinder "NetworKit::IndependentSetFinder":
+		_IndependentSetFinder() except +
+		vector[bool_t] run(const _Graph& G) except +
+		bool_t isIndependentSet(const vector[bool_t]& nodeSet, const _Graph& G)
 
+cdef class IndependentSetFinder:
+	""" 
+	IndependentSetFinder()
+	
+	Abstract base class for independent set algorithms.
+	"""
+	cdef _IndependentSetFinder* _this
 
-# FIXME: check correctness
-cdef class Luby:
-	""" Luby's parallel maximal independent set algorithm"""
-	cdef _Luby _this
+	def __cinit__(self, *args):
+		# The construction is handled by the subclasses
+		return
 
-	def run(self, Graph G not None):
-		""" Returns a bool vector of length n where vec[v] is True iff v is in the independent sets.
-		
-		Parameters:
-		-----------
+	def __dealloc__(self):
+		if self._this is not NULL:
+			del self._this
+			self._this = NULL
+
+	def run(self, Graph G):
+		"""
+		run(G)
+
+		Returns a list, where an entry indicates whether node i is part of an independent set.
+
+		Parameters
+		----------
 		G : networkit.Graph
 			The graph.
-		
-		Returns:
-		--------
-		vector
-			A bool vector of length n.
+
+		Returns
+		-------
+		list(bool)
+			A list of bools.
 		"""
 		return self._this.run(G._this)
-		# TODO: return self
+
+	def isIndependentSet(self, nodeSet, Graph G):
+		""" 
+		isIndependentSet(set, G)
+		
+		Checks whether a given set is an independent set in graph G.
+
+		Parameters
+		----------
+		G : networkit.Graph
+			The input graph.
+		nodeSet : list(bool)
+			Input list, where each entry represents a node. If entry is True, the node is part of the set; False if not.
+
+		Returns
+		-------
+		bool
+
+		"""
+		return self._this.isIndependentSet(nodeSet, G._this)
+
+cdef extern from "<networkit/independentset/Luby.hpp>":
+
+	cdef cppclass _Luby "NetworKit::Luby"(_IndependentSetFinder):
+		_Luby() except +
+
+cdef class Luby(IndependentSetFinder):
+	""" 
+	Luby()
+
+	Luby's parallel maximal independent set algorithm.
+	"""
+
+	def __cinit__(self, *args):
+		self._this = new _Luby()
+
+	def __dealloc__(self):
+		if self._this is not NULL:
+			del self._this
+			self._this = NULL
```

Comment:
This probably leads to segfault due to a double free (`__dealloc__` is already implemented in `IndependentSetFinder`). I would recommend writing a simple Python test for this class.

Reply:
I think the double free is somehow avoided, at least it doesn't happen locally. I removed it nevertheless and added a small test. Should we do the same for linkprediction?

Reply:
(More a general question/remark ... since only a few modules are covered by Python tests).

Reply:
I think that we should have simple tests for the Python frontend. However, this should be done in separate PRs.